### PR TITLE
feat: providing granuarity on network ACLs

### PR DESCRIPTION
**Pull request by imprateeksh**
### Description

Narrowing down the network connectivity rules as mentioned below:
 - **On inbound**: source : VPC address prefixes ; destination cidr 10.0.0.0/8 
 - **On outbound**: source cidr is 10.0.0.0/8 , destination: VPC address prefixes. 
 (There is one ACL rule per VPC address prefix.)

#### No release required

- [x] Examples or tests (addition or updates of examples or tests)
- [ ] Documentation update
- [ ] CI-related update (pipeline, etc.)
- [ ] Other changes that don't affect Terraform code

#### Release required

- [ ] Bug fix (patch release (`x.x.X`): Change that fixes an issue and is compatible with earlier versions)
- [x] New feature (minor release (`x.X.x`): Change that adds functionality and is compatible with earlier versions)
- [ ] Breaking change (major release (`X.x.x`): Change that is likely incompatible with previous versions)

##### Release notes content

Reducing the ACL network rules by narrowing down the network connectivity as mentioned below:
 - **On inbound**: source : VPC address prefixes ; destination cidr 10.0.0.0/8 
 - **On outbound**: source cidr is 10.0.0.0/8 , destination: VPC address prefixes. 
 (There is one ACL rule per VPC address prefix.)

---

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- Merge by using "Squash and merge".
- Use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author.

    The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).

### CODE REVIEW

File: network_acls.tf [hunk 1]
```diff
@@ -46,31 +46,34 @@ locals {
       icmp        = null
     }
   ]
+  vpc_inbound_rule = [
+    for zone, addresses in var.address_prefixes : [for index, address in addresses :
```
Comment (vburckhardt) [inline network_acls.tf:50]: I think the loop should be based on https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/data-sources/is_vpc_address_prefixes . var.address_prefixes is optional. In the case where it is not specified the address prefixes are determined by the VPC stack, so in any case using the data source to retrieve the actual prefixes from the server is needed.

**Review by vburckhardt — CHANGES_REQUESTED** [review]

File: network_acls.tf (hunk 1 above)
Comment (imprateeksh) [reply network_acls.tf]: Sure @vburckhardt. I have made the changes to use data source `is_vpc_address_prefix` always for getting address_prefixes.
Pipeline is failing, not sure at the moment what is going wrong here.

File: network_acls.tf [hunk 2]
```diff
@@ -2,6 +2,10 @@
 # Network ACL
 ##############################################################################
 
+data "ibm_is_vpc_address_prefixes" "get_address_prefixes" {
```
[the remark this answers is not in the document]
Comment (imprateeksh) [reply network_acls.tf:5]: Sure @rajatagarwal-ibm - I have added the cidr blocks as the output from address prefixes.

**Comment by vbontempi** [comment]
For a more fined setup of outbound and inbound acl rules, should we limit outbound destination and inbound source to the vpc prefixes in the same way it is done for outbound source and inbound destination?

We are currently setting to the network cidr or to everything if this is null

**Comment by imprateeksh** [comment]
> 

Hi @vbontempi Thank you for providing your valuable input. I think we need to restrict the destination in case of inbound and source in case of outbound to `10.0.0.0/8` (_for now atleast_)  as @vburckhardt mentioned in the issue [here](https://github.com/terraform-ibm-modules/terraform-ibm-landing-zone-vpc/issues/357) , also the default value of variable network_cidr is set to `10.0.0.0/8` so this I think will not be `null` but this is generic null check which already exists.
image

File: network_acls.tf [hunk 3]
```diff
@@ -47,31 +47,35 @@ locals {
...
   ]
 
-  vpc_connectivity_rules = [
-    # All connectivity across any subnet within VPC
-    # TODO: narrow down to VPC address spaces
+  vpc_inbound_rule = [
+    for address in data.ibm_is_vpc_address_prefixes.get_address_prefixes.address_prefixes :
     {
-      name        = "ibmflow-allow-vpc-connectivity-inbound"
+      name        = "ibmflow-allow-vpc-connectivity-inbound-${substr(address.id, -4, -1)}" # Providing unique rule names
```
[the remark this answers is not in the document]
Comment (imprateeksh) [reply network_acls.tf:53]: @rajatagarwal-ibm Thank you for your input here. As per your suggestion I initially thought of adding a random string but we have a length validation too i.e. `1<= acl_name<=63`
Using the default names, the length is 38/39, so I thought of keeping it short by adding last 4 digits of random ids of address suffixes.
image

But if you suggest, random string (of length around 15/16 chars) appended at last should be a good choice, I will go ahead and make the change.

[the remark this answers is not in the document]
Comment (rajatagarwal-ibm) [reply network_acls.tf:53]: So a sample ID would be "r006-88ce12bd-5611-48a5-84b0-dc55c640d634" and you are taking the last 4 chars for it. I think that should be ok

[the remark this answers is not in the document]
Comment (imprateeksh) [reply network_acls.tf:53]: sure.

File: main.tf [hunk 4]
```diff
@@ -36,6 +36,9 @@ resource "ibm_is_vpc_address_prefix" "address_prefixes" {
   cidr     = each.value.cidr
 }
 
+data "ibm_is_vpc_address_prefixes" "get_address_prefixes" {
```
Comment (vburckhardt) [inline main.tf:39]: I suspect that this data source is going to return immediately after the vpc is created, but before any of the ibm_is_vpc_address_prefix resource block are executed. In such case, the result in data source will be missed the prefixes address through the resource. Either add a dependency here, or lookup in the ibm_is_vpc_address_prefix resource if it exists (in the logic at address in data.ibm_is_vpc_address_prefixes.get_address_prefixes.address_prefixes )

**Review by vburckhardt — CHANGES_REQUESTED** [review]
Hello @imprateeksh - just one comment here.

**Comment by vburckhardt** [comment]
@vbontempi - I think the source for outbound and destination for inbound are not really relevant here anyways. In outbound, the source is always going to be the subnets the acl is attached to. And inbound, the destination is always the subnet to acl is attached to. So whether we put 10.0.0.0/8 (if the subnets fall in this address prefix) or 0.0.0.0/0 we end up with same behavior here.

**Comment by imprateeksh** [comment]
@vburckhardt / @ocofaigh  - Could you please approve/merge this PR? I have made the changes as suggested.

**Comment by vburckhardt** [comment]
Missing , ibm_is_vpc_address_prefix.subnet_prefix as well in the depends_on list - I added it in a commit to this PR so that we can go ahead and merge it. Without this depends on, the acl is missing the prefixes added at https://github.com/terraform-ibm-modules/terraform-ibm-landing-zone-vpc/blob/1e89ae2dc5c50ceb40f0109b50edecdaedd5a569/subnet.tf#L16

**Review by vburckhardt — APPROVED** [review]

**Comment by imprateeksh** [comment]
Thank you @vburckhardt

**Pull request merged**